pubsub/topic.go

```diff
@@ -528,13 +538,43 @@ func (t *Topic) initBundler() {
 	}
 	t.scheduler.BundleByteThreshold = t.PublishSettings.ByteThreshold
 
+	fcs := DefaultPublishSettings.FlowControlSettings
+	// FlowControlSettings.MaxOutstandingBytes and BufferedByteLimit have generally
+	// the same behavior, with the latter always returning an error. BufferedByteLimit
+	// is deprecated in favor of MaxOutstandingBytes.
+	//
+	// While we continue to support both, check if either has been set directly, and use
+	// that value. If both have been set, we will respect MaxOutstandingBytes first.
+	var mo, bb bool
+	if t.PublishSettings.BufferedByteLimit > 0 {
+		bb = true
+	}
+	if t.PublishSettings.FlowControlSettings.MaxOutstandingBytes > 0 {
+		mo = true
+	}
+	if bb && !mo {
+		fcs.MaxOutstandingBytes = t.PublishSettings.BufferedByteLimit
+	} else if mo {
+		fcs.MaxOutstandingBytes = t.PublishSettings.FlowControlSettings.MaxOutstandingBytes
+	}
+	if t.PublishSettings.FlowControlSettings.LimitExceededBehavior != FlowControlBlock {
+		fcs.LimitExceededBehavior = t.PublishSettings.FlowControlSettings.LimitExceededBehavior
+	}
+	if t.PublishSettings.FlowControlSettings.MaxOutstandingMessages > 0 {
+		fcs.MaxOutstandingMessages = t.PublishSettings.FlowControlSettings.MaxOutstandingMessages
+	}
+
+	t.flowController = newFlowController(fcs)
+
 	bufferedByteLimit := DefaultPublishSettings.BufferedByteLimit
 	if t.PublishSettings.BufferedByteLimit > 0 {
 		bufferedByteLimit = t.PublishSettings.BufferedByteLimit
 	}
 	t.scheduler.BufferedByteLimit = bufferedByteLimit
 
-	t.scheduler.BundleByteLimit = MaxPublishRequestBytes - calcFieldSizeString(t.name)
+	// Calculate the max limit of a single bundle. 5 comes from the number of bytes
+	// needed to be reserved for encoding the PubsubMessage repeated field.
+	t.scheduler.BundleByteLimit = MaxPublishRequestBytes - calcFieldSizeString(t.name) - 5
```

Comment:
Is the 5 an implementation detail of proto? Is removing 5 related to the change to add the new form of flow control or an unrelated bug fix?

Reply:
It's kind of an unrelated bug that cropped up when I was trying to implement the tests. In `topic.Publish`, the additional size of each message was incorrectly calculated as a `PublishRequest` with a single message. Now, it's correctly calculated just the size of a `PubsubMessage`, and the number 5 accounts for the reservation of the array/repeated field instead.